pkg/mutation/mutators/core/setter.go

```diff
@@ -0,0 +1,166 @@
+	// field on that object that should be mutated. It is up to the
+	// implementor to actually mutate the object.
+	SetValue(obj map[string]interface{}, key string) error
+
+	// KeyedListOkay returns whether this setter can handle keyed lists.
+	// If it can't, an attempt to mutate a keyed-list-type field will
+	// result in an error.
+	KeyedListOkay() bool
+
+	// KeyedListValue is the value that will be assigned to the
+	// targeted keyed list entry. Unlike SetValue(), this does
+	// not do mutation directly.
+	KeyedListValue() (map[string]interface{}, error)
+
+	// MergeListElements tries to merge the old list element
+	// with the new list element.
+	MergeListElements(old interface{}, new map[string]interface{}) (map[string]interface{}, error)
+}
+
+var _ Setter = &defaultSetter{}
+
+func NewDefaultSetter(value interface{}) Setter {
+	return &defaultSetter{
+		value: value,
+	}
+}
+
+// defaultSetter is the default implementation of the Setter interface that supports
+// assigning plain values and external data placeholders.
+type defaultSetter struct {
+	value interface{}
+}
+
+func (s *defaultSetter) KeyedListOkay() bool { return true }
+
+func (s *defaultSetter) KeyedListValue() (map[string]interface{}, error) {
+	valueAsObject, ok := s.value.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("assign.value for keyed list is not an object: %+v", s.value)
+	}
+	return valueAsObject, nil
+}
+
+func (s *defaultSetter) SetValue(obj map[string]interface{}, key string) error {
+	value := s.value
+	incomingPlaceholder, isIncomingPlaceholder := types.ToExternalDataPlaceholder(value)
+	if _, ok := obj[key]; ok {
+		switch prev := obj[key].(type) {
+		case map[string]interface{}:
+			if previousPlaceholder, isPreviousPlaceholder := types.ToExternalDataPlaceholder(prev); isPreviousPlaceholder {
+				// replace old placeholder with new one
+				if isIncomingPlaceholder {
+					incomingPlaceholder.ValueAtLocation = previousPlaceholder.ValueAtLocation
+				} else if valueAsString, err := encodeToString(value); err == nil {
+					// assign incoming value as string to the previous placeholder
+					incomingPlaceholder = previousPlaceholder
+					incomingPlaceholder.ValueAtLocation = valueAsString
+				} else {
+					// return an error if the incoming value can't be encoded to a string
+					return err
+				}
+			}
+		default:
+			if isIncomingPlaceholder {
+				valueAsString, err := encodeToString(prev)
+				if err != nil {
+					return err
+				}
+				incomingPlaceholder.ValueAtLocation = valueAsString
+			}
+		}
+	}
+
+	var err error
+	if incomingPlaceholder != nil {
+		value, err = incomingPlaceholder.ToUnstructured()
+		if err != nil {
+			return fmt.Errorf("unable to convert external data placeholder to unstructured: %w", err)
+		}
+	}
+
+	obj[key] = value
+	return nil
+}
+
+func (s *defaultSetter) MergeListElements(old interface{}, new map[string]interface{}) (map[string]interface{}, error) {
```

Comment:
Is this function necessary, given that we can't have external data reference list elements directly?

Reply:
It's correct that we do not allow external data reference list elements but we allow external data reference as the field of a list element (e.g. `spec.containers[name:*].image`). 

Let's say we have the following two mutations:
1. [external data assign] convert image tag to digest for `spec.containers[name:*].image`
1. [non-external data assign] sidecar injection to `spec.containers[name:busybox]`

Here is what it will look like without the above function:

iteration 0:
```yaml
spec:
  containers:
    - name: nginx
      image: nginx:latest
```

iteration 1:
```yaml
spec:
  containers:
    - name: nginx
      image: Placeholder{ValueAtLocation:"nginx:latest"}
    - name: busybox
      image: busybox
```

iteration 1.5:
```yaml
spec:
  containers:
    - name: nginx
      image: Placeholder{ValueAtLocation:"nginx"}
    - name: busybox
      image: Placeholder{ValueAtLocation:"busybox"}
```

iteration 2:
```yaml
spec:
  containers:
    - name: nginx
      image: Placeholder{ValueAtLocation:"nginx"}
    - name: busybox
      image: busybox
```

^ current behavior is to overwrite list element once we found it - https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/mutation/mutators/core/mutation_function.go#L242

MergeListElements will handle the merging of placeholder fields and nonplaceholder fields. In the example above, we would need to merge the image field - `Placeholder{ValueAtLocation:"busybox"}` and `busybox`.

Reply:
It's subtle, but overwriting is the correct behavior here and is expected.

The issue here is consistency, mutators should behave roughly the same whether we're specifying constant values or external data providers.

If I have:

   1. [assign] `spec.containers[name:*].image` = "foobar"
   2. [assign] sidecar injection to `spec.containers[name:busybox]`
   
mutator (1) would be clobbered, we should preserve this behavior for external-data-based values as well.

Clobbering one mutator with another one is also a valid use-case, suppose an admin specifically wants to avoid calling the external data mutator for the sidecar? Syntactically we have two ways of providing this functionality:

  1. Allow the admin to specify the sidecar mutator as occurring after the external data mutator OR...
  2. Allow the admin to specify some syntax that says "ignore X entries", such as: `spec.containers[name:*, exceptions:["busybox"]].image` (or however we create the syntax)
  
(2) has a few important limitations:

   - It does not handle the addition of new "ignore this sidecar" mutators, the original mutator will need to be updated to handle every new exception
   - It cannot differentiate between user-provided state and state injected by other mutators. So, if a user specified their own sidecar `{name: "busybox", image: "exempt-from-retagging"}` in a namespace where we don't add a busybox sidecar, then (2) would exempt the image from retagging, (1) would not.

IMO, the edge case you highlight is a strong argument for why we need to give the user some control over the ordering of mutations.

Also, note that to fully implement this merge logic, we would need to recurse through the individual object during the merge to avoid clobbering in cases like:

   1. [external data assign] convert image tag to digest for spec.containers[name:*].some.interior.field
   2. [non-external data assign] sidecar injection to spec.containers[name:busybox]
   
At that point, we're recursing within a recursion, which could get pretty expensive, complexity-wise (think O(N^N) worst-case).

Reply:
An alternative to user-provided mutation order would be to sort mutators by `location` depth, with least-deep applied first. This would minimize the number of clobbered mutations. It would not be incompatible with adding user-provided sort order clues later.

Reply:
Got it. To minimize the scope of the PR, I will just preserve the original behavior for external-data-based values. We can definitely follow up on this (we already have https://github.com/open-policy-agent/gatekeeper/issues/1133 to track the issue)